File: Graph/Python/Build a Matrix With Conditions.py
# Problem link: https://leetcode.com/problems/build-a-matrix-with-conditions/description/
# Time: O(k * k)
# Space: O(k * k)

# Approach: DFS
class Solution:
    def buildMatrix(self, k: int, rowConditions: List[List[int]], colConditions: List[List[int]]) -> List[List[int]]:

        rowOrder = self.topoSort(rowConditions, k)
        colOrder = self.topoSort(colConditions, k)
        
        if not rowOrder or not colOrder:
            return []
        
        result = [[0] * k for _ in range(k)]
        rowIdx = {val : idx for idx, val in enumerate(rowOrder)}
        colIdx = {val : idx for idx, val in enumerate(colOrder)}

        for idx in range(1, k + 1):
            result[rowIdx[idx]][colIdx[idx]] = idx
        
        return result
    
    def topoSort(self, edges, num):
        adjList = defaultdict(list)

        for u, v in edges:
            adjList[u].append(v)
     
        path, visited, result = [0] * (num + 1), [0] * (num + 1), []

        for idx in range(1, num + 1):
            if self.dfs(adjList, path, visited, result, idx):
                return []
        
        return result[::-1]
    
    def dfs(self, adjList, path, visited, result, node):
        if path[node]:
            return True
        elif visited[node]:
            return False
        
        path[node] = visited[node] = 1

        for ver in adjList[node]:
            if self.dfs(adjList, path, visited, result, ver):
                return True

        path[node] = 0
        result.append(node)

        return False

# Approach: BFS
class Solution:
    def buildMatrix(self, k: int, rowConditions: List[List[int]], colConditions: List[List[int]]) -> List[List[int]]:
        rowIndex = self.topoSort(rowConditions, k)
        colIndex = self.topoSort(colConditions, k)

        if len(rowIndex) != k or len(colIndex) != k:
            return []

        result = [[0] * k for _ in range(k)]
        indicesDetails = {}

        for idx in range(len(rowIndex)):
            indicesDetails[rowIndex[idx]] = idx

        for idx in range(len(colIndex)):
            result[indicesDetails[colIndex[idx]]][idx] = colIndex[idx]
        
        return result
    

    def topoSort(self, edges, totalNodes):
        indegree, topoOrder = [0] * (totalNodes), []
        adjList = {}
        
        for u, v in edges:
            if u not in adjList:
                adjList[u] = []
            adjList[u].append(v)
            indegree[v - 1] += 1
        
        queue, visit = deque(), set()
        
        for idx in range(len(indegree)):
            if indegree[idx] == 0:
                queue.append(idx + 1)
        
        while queue:
            node = queue.popleft()
            topoOrder.append(node)
            
            if node in adjList:
                for ver in adjList[node]:
                    indegree[ver - 1] -= 1
                    if indegree[ver - 1] == 0:
                        queue.append(ver)
        
        return topoOrder

File: Graph/Java/Build a Matrix With Conditions.java
// Problem link: https://leetcode.com/problems/build-a-matrix-with-conditions/description/
// Time: O(k * k)
// Space: O(k * k)

class Solution {
    public int[][] buildMatrix(int k, int[][] rowConditions, int[][] colConditions) {
        
        int[] rowOrder = this.topoSort(rowConditions, k);
        int[] colOrder = this.topoSort(colConditions, k);

        if(rowOrder.length == 0 || colOrder.length == 0) {
            return new int[][] {};
        }

        int[][] result = new int[k][k];
        int[] rowIdx = new int[k + 1], colIdx = new int[k + 1];

        for(int idx = 0; idx < k; idx++) {
            rowIdx[rowOrder[idx]] = idx;
            colIdx[colOrder[idx]] = idx;
        }

        for(int idx = 1; idx <= k; idx++) {
            result[rowIdx[idx]][colIdx[idx]] = idx;
        }

        return result;
    }

    private int[] topoSort(int[][] edges, int num) {
        Map<Integer, List<Integer>> adjList = new HashMap<>();
        int[] visited = new int[num + 1], path = new int[num + 1], result = new int[num + 1];
        result[num] = num - 1;

        for(int[] pair : edges) {
            adjList.putIfAbsent(pair[0], new ArrayList<>());
            adjList.get(pair[0]).add(pair[1]);
        }

        for(int idx = 1; idx <= num; idx++) {
            if(this.dfs(adjList, visited, path, result, idx)) {
                return new int[] {};
            }
        }

        return result;
    }

    private boolean dfs(Map<Integer, List<Integer>> adjList, int[] visited, int[] path, int[] result, int node) {
        if(path[node] == 1) {
            return true;
        }
        else if(visited[node] == 1) {
            return false;
        }

        path[node] = visited[node] = 1;

        if(adjList.containsKey(node)) {
            for(int ver : adjList.get(node)) {
                if(this.dfs(adjList, visited, path, result, ver)) {
                    return true;
                }
            }
        }

        path[node] = 0;
        result[result[result.length - 1]--] = node;

        return false;
    }
}
